fairscale/nn/data_parallel/fully_sharded_data_parallel.py

```diff
@@ -1721,35 +1721,50 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
         # reductions in post_backward stream.
         self._streams["post_backward"].wait_stream(torch.cuda.current_stream())
         with torch.cuda.stream(self._streams["post_backward"]):
-            orig_grad_data = param.grad.data
+            if param.main_grad is not None:
+                orig_grad_data = param.main_grad
+            else:
+                orig_grad_data = param.grad
 
             if self.fp32_reduce_scatter:
-                # Cast grad to FP32.
-                param.grad.data = param.grad.data.float()
+                if param.grad is not None:
+                    if param.main_grad is not None:
+                        param.main_grad.copy_(param.grad.float())
```

Comment:
nit: torch can upcast and copy in one kernel:
```suggestion
                        param.main_grad.copy_(param.grad)
```

<details>
<summary> Existing </summary>

Upcast kernel + copy kernel
![Screenshot 2023-10-04 at 12 27 52 PM](https://github.com/facebookresearch/fairscale/assets/31054793/2102f0ee-6e67-4a3f-9af6-fd1af69fd619)

</details>

<details>
<summary> New </summary>

Only upcast kernel

![Screenshot 2023-10-04 at 12 29 32 PM](https://github.com/facebookresearch/fairscale/assets/31054793/c8e977cb-f75b-4c6d-ad4a-cfb3b7e44137)


</details>


<details>
<summary> Correctness Example </summary>

```
>>> t_fp32 = torch.empty((4,))
>>> t_bf16 = torch.randn((4,), dtype=torch.bfloat16)
>>> t_fp32
tensor([-8.3762e-20,  3.0801e-41, -1.3043e-16,  3.0801e-41])
>>> t_bf16
tensor([-1.3516, -0.5156, -0.6055,  0.3535], dtype=torch.bfloat16)
>>> t_fp32.copy_(t_bf16)
tensor([-1.3516, -0.5156, -0.6055,  0.3535])
>>> t_fp32
tensor([-1.3516, -0.5156, -0.6055,  0.3535])
```

</details>